File: ShadowFrameworkLite20/src/shadow/renderer/data/proxies/SFClonedArrayReference.java
package shadow.renderer.data.proxies;

import shadow.renderer.SFAbstractReferenceNode;
import shadow.renderer.SFNode;
import shadow.renderer.SFReferenceNode;
import shadow.renderer.data.SFDataAsset;
import shadow.renderer.data.SFDataAssetList;
import shadow.renderer.data.SFDataObjectsList;
import shadow.renderer.data.SFLibraryReference;
import shadow.system.SFInitiable;
import shadow.system.data.SFNamedParametersObject;
import shadow.system.data.objects.SFString;

public class SFClonedArrayReference extends SFDataAsset<SFNode>{

	//private SFIndexedProxyDataObject mappedProxy=new SFIndexedProxyDataObject();
	private SFDataObjectsList<SFString> properties=new SFDataObjectsList<SFString>(new SFString());
	private SFDataObjectsList<SFString> constants=new SFDataObjectsList<SFString>(new SFString());
	private SFDataAssetList<SFInitiable> elements=new SFDataAssetList<SFInitiable>();
	private SFLibraryReference<SFNode> node=new SFLibraryReference<SFNode>(null);
	private SFIndexedProxyDataCenter proxyDataCenter=new SFIndexedProxyDataCenter();
	
	
	public SFClonedArrayReference() {
		SFNamedParametersObject parameters=new SFNamedParametersObject();
		parameters.addObject("node", node);
		parameters.addObject("properties", properties);
		parameters.addObject("constants", constants);
		parameters.addObject("elements", elements);
		//parameters.addObject("mappedProxy", mappedProxy);
		setData(parameters);
	}
	
	public void setNodeModel(SFDataAsset<SFNode> model){
		this.node.setDataset(model);
	}

	public SFLibraryReference<SFNode> getNode() {
		return node;
	}
	
	public void addName(String name){
		properties.add(new SFString(name));
	}
	
	@SuppressWarnings("unchecked")
	public void setData(SFDataAsset<?> dataset)
			throws RuntimeException {
		elements.add((SFDataAsset<SFInitiable>)dataset);
	}
	
	public SFIndexedProxyDataCenter getProxyDataCenter() {
		return proxyDataCenter;
	}

	@Override
	protected SFNode buildResource() {
		SFAbstractReferenceNode node=new SFReferenceNode();
		
		int index=0;
		for (int i = 0; i < elements.size(); ) {
			for (int j = 0; j < properties.size(); j++,i++) {
				proxyDataCenter.setData(properties.get(j).getString(), index, elements.get(i));
			}
			index++;
		}
		

		for (int i = 0; i < constants.size(); i++) {
			proxyDataCenter.addConstant(constants.get(i).getString());
		}
		
		proxyDataCenter.setup();

		setUpdateMode(true);
		
		for (int i = 0; i < proxyDataCenter.size(); i++) {
			
			proxyDataCenter.setIndex(i);
			
			SFNode model=this.node.getDataset().getResource();
			
			node.addNode(model);
		}
		
		setUpdateMode(false);

		proxyDataCenter.unset();
		
		return node;
	}
	
}

File: ShadowFrameworkJS/src/shadow/renderer/data/proxies/SFClonedArrayReference.js


function SFClonedArrayReference(){
	this.properties=new SFDataObjectsList(new SFString());
	this.constants=new SFDataObjectsList(new SFString());
	this.elements=new SFDataAssetList();
	this.node=new SFLibraryReference(null);
	this.proxyDataCenter=new SFIndexedProxyDataCenter();

	var parameters=new SFNamedParametersObject();
	parameters.addObject(this.node);
	parameters.addObject(this.properties);
	parameters.addObject(this.constants);
	parameters.addObject(this.elements);
	//parameters.addObject("mappedProxy", mappedProxy);
	this.setData(parameters);
}

inherit(SFClonedArrayReference,SFDataAsset);
	
SFClonedArrayReference.prototype["buildResource"]=function(){
	var node=new SFReferenceNode();
		
	var index=0;
	for (var i = 0; i < this.elements.size(); ) {
		for (var j = 0; j < this.properties.size(); j++,i++) {
			this.proxyDataCenter.setData(this.properties.get(j).getString(), index, this.elements.get(i));
		}
		index++;
	}

	for (var i = 0; i < this.constants.size(); i++) {
		this.proxyDataCenter.addConstant(this.constants.get(i).getString());
	}
	
	this.proxyDataCenter.setup();

	SFDataAsset_setUpdateMode(true);
	
	for (var i = 0; i < this.proxyDataCenter.size; i++) {
		
		this.proxyDataCenter.setIndex(i);
		
		var model=this.node.getDataset().getResource();
		
		node.addNode(model);
	}
	
	SFDataAsset_setUpdateMode(false);

	this.proxyDataCenter.unset();
	
	return node;
};
	

SFClonedArrayReference.prototype["generateNewDatasetInstance"]=function(){
	return new SFClonedArrayReference();
};

